sklearn/tree/_splitter.pyx

```diff
@@ -43,13 +44,19 @@ cdef DTYPE_t FEATURE_THRESHOLD = 1e-7
 # in SparseSplitter
 cdef DTYPE_t EXTRACT_NNZ_SWITCH = 0.1
 
-cdef inline void _init_split(SplitRecord* self, SIZE_t start_pos) nogil:
-    self.impurity_left = INFINITY
-    self.impurity_right = INFINITY
-    self.pos = start_pos
-    self.feature = 0
-    self.threshold = 0.
-    self.improvement = -INFINITY
+cdef inline void _init_split(SplitRecord* split, SIZE_t start_pos, SIZE_t n) nogil:
+    split.feature = 0
+    split.threshold = 0.
+    split.pos = start_pos
+    split.improvement = -INFINITY
+    split.impurity_left = INFINITY
+    split.impurity_right = INFINITY
+    split.weighted_n_left = 0.
+    split.weighted_n_right = 0.
+    split.sum_left = <double*> calloc(n, sizeof(double))
+    split.sum_right = <double*> calloc(n, sizeof(double))
```

Comment:
`safe_realloc` requires the gil, so I'd have to release the gil at each step. `calloc` does not require the gil. What do you think is the best choice here?